### ISSUE TITLE
Object undefined checks prior to usage

### ISSUE THREAD

**Issue by bgp1** [eslint/eslint#13941]
**Please describe what the rule should do:**
Similar to a null reference check in C# this rule should enforce checking for `undefined` prior to usage of an object property. Note this should *not* attempt to check if a property exists but rather confirm that a check has been made to determine the parent object is not `undefined`.

**What new ECMAScript feature does this rule relate to?**

**What category of rule is this? (place an "X" next to just one item)**

[X] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

Consider the following example:
```js
console.log(data.event.title);
```
This would be flagged because you're utilizing both `event` object and `title` property without checking to see if `data` is defined and `event` is defined. The following would be a correct utilization:
```js
if (typeof data !== 'undefined' && typeof data.event !== 'undefined') {
  console.log(data.event.title);
}
```

**Why should this rule be included in ESLint (instead of a plugin)?**
It helps promote clean and usable code and adding undefined checks improves the user experience. Enforcing undefined checks in an organization in a manual PR check fashion can work but is not failsafe without this rule. 

**Are you willing to submit a pull request to implement this rule?**
Potentially

**Comment by mdjermanovic** [comment]
Hi @bgp1, thanks for the issue!

Per our [rules policies](https://eslint.org/blog/2020/05/changes-to-rules-policies), new rules have to be related to ECMAScript features added within the last 12 months, so this proposal doesn't qualify.

Also, I don't think it's common to _always_ check if `foo` is defined before accessing `foo.bar`, or if `foo.bar` is defined before accessing `foo.bar.baz`. If they're expected to be defined at some point, adding these checks could hide bugs and make it difficult to find the root cause.

**Comment by nzakas** [comment]
As @mdjermanovic mentioned, this proposal doesn't qualify for inclusion as a core rule in ESLint. However,  you can always create your own custom rules to use with ESLint, and that's what I'd suggest you do here. Thanks!
